### PR TITLE
Extend mixpanel's definitions

### PR DESCRIPTION
**Pull request by rasom**
is necessary for https://github.com/status-im/status-react/issues/3792

`mixpanel_events.edn` is now cljs file, so that we can define functions there if necessary. Each event may contain `:filter-fn` and `:data-fn` options. The former one allows to filter event using its parameters, the latter can return map which will be merged with `:properties` before sending.

Also `matching-events` function was rewritten so that it doesn't iterate over all possible events on each re-frame's events. It took >2ms for each re-frame event on slow devices, 10-12 times faster now. 

 
status: ready
#

### CODE REVIEW

**Comment by jeluard** [comment]
Doesn't look like `filter-fn` is used at all. Do you have a specific scenario in mind where it can be useful?

**Review by janherich — APPROVED** [review]

File: src/status_im/utils/mixpanel.cljs [hunk 1]
```diff
@@ -75,20 +76,28 @@
...
-        :else
-        (= event trigger)))
+(defn matching-events [[event-name first-arg :as event] triggers]
+  (let [cnt      (count event)
+        triggers (cond->
+                  (get-in triggers [event-name event-tag])
 
-(defn matching-events [event definitions]
-  (reduce-kv #(if (matches? event %2) (conj %1 %3) %1) [] definitions))
+                  (>= 2 cnt)
```
Comment (jeluard) [inline src/status_im/utils/mixpanel.cljs]: Can you comment the new matching logic?

File: src/status_im/utils/mixpanel.cljs [hunk 2]
```diff
@@ -75,20 +76,28 @@
...
+                  (concat (get-in triggers [event-name first-arg event-tag]))
 
+                  (> 2 cnt)
+                  (concat (get-in triggers (conj [event] event-tag))))]
```
Comment (cammellos) [inline src/status_im/utils/mixpanel.cljs]: Is this not just `[event event-tag]` rather than `(conj [event] event-tag)` ?

**Review by cammellos — APPROVED** [review]

File: src/status_im/utils/mixpanel.cljs (hunk 2 above)
Comment (rasom) [reply src/status_im/utils/mixpanel.cljs]: you are right, will fix soon

**Comment by rasom** [comment]
@jeluard it's not used in this PR, i just wanted to split the big one for #3792 into smaller PRs
https://github.com/status-im/status-react/pull/3969

File: src/status_im/utils/mixpanel.cljs (hunk 2 above)
Comment (rasom) [reply src/status_im/utils/mixpanel.cljs]: wait it should be `(conj event event-tag)`. I will add a test case

**Comment by rasom** [comment]
@jeluard @cammellos please review changes

**Review by jeluard — APPROVED** [review]

**Review by cammellos — APPROVED** [review]

**Comment by lukaszfryc** [comment]
Looks good 👍 

I checked sending Mixpanel events while being offline and online

**Pull request merged**